File: InfiniteHorizonLQR.m
% stochastic optimal control example
% infinite horizon LQR

% problem data
n = 12;
m = 4;

A = randn(n,n); A = A/max(abs(eig(A)));
B = randn(n,m);
% rank(ctrb(A,B)) % check for controllability 

Q = eye(n);
R = 1*eye(m);

W = 1*eye(n);

%% optimal closed-loop feedback policy
P = zeros(n,n); % initialize
Pplus = randn(n,n); % data structure for optimal cost udpate


% value iteration recursion
t = 0;
while norm(P - Pplus, 'fro') > 1e-6
    t = t + 1
    P = Pplus;
    Pplus = Q + A'*P*A - A'*P*B*inv(R + B'*P*B)*B'*P*A;
end

K = -inv(R + B'*P*B)*B'*P*A;

% check against built in ARE solver
[P1, K1, L1] = idare(A,B,Q,R);

norm(P - P1, 'fro')
norm(K + K1, 'fro')

%% Monte Carlo simulation with Ns noise sequence samples
Ns = 100;
T = 100;
x0 = 40*randn(n,1);
x = zeros(n,T+1,Ns);
u = zeros(m,T,Ns);
J = zeros(T+1,Ns);
J(1,:) = x0'*x0;
w = sqrt(1)*randn(n,T,Ns);

for i=1:Ns
    x(:,1,i) = x0;
    for j=2:T+1
        u(:,j-1,i) = K*x(:,j-1,i);
        x(:,j,i) = (A + B*K)*x(:,j-1,i) + w(:,j-1,i);
        J(j,i) = x(:,j,i)'*Q*x(:,j,i) + u(:,j-1,i)'*R*u(:,j-1,i);
    end
end

% plot sample trajectory
figure;
stairs(0:T,x(:,:,1)','LineWidth',2)
xlabel('time');
ylabel('states');
title('sample state trajectory');
set(gca, 'fontsize', 24);

figure;
stairs(0:T-1,u(:,:,1)','LineWidth',2)
xlabel('time');
ylabel('inputs');
title('sample input trajectory');
set(gca, 'fontsize', 24);

File: InfiniteHorizonLQR.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.linalg import solve_discrete_are

# Problem data
n = 15
m = 4

A = np.random.randn(n, n)
B = np.random.randn(n, m)

# Check for controllability
is_controllable = np.linalg.matrix_rank(np.hstack((B, A @ B, A @ A @ B, A @ A @ A @ B))) == n
print("Controllability Check:", is_controllable)

Q = np.eye(n)
R = np.eye(m)

W = np.eye(n)

# Optimal closed-loop feedback policy using value iteration
P = np.zeros((n, n))
Pplus = np.random.randn(n, n)

# Value iteration recursion
t = 0
while np.linalg.norm(P - Pplus, 'fro') > 1e-6:
    t += 1
    P = Pplus
    Pplus = Q + A.T @ P @ A - A.T @ P @ B @ np.linalg.inv(R + B.T @ P @ B) @ B.T @ P @ A

K = -np.linalg.inv(R + B.T @ P @ B) @ B.T @ P @ A


# compare with built-in scipy function
P1 = solve_discrete_are(A, B, Q, R)
K1 = -np.linalg.inv(R + B.T @ P1 @ B) @ B.T @ P1 @ A

# Compare P and P1, K and K1
norm_P_diff = np.linalg.norm(P - P1, 'fro')
norm_K_diff = np.linalg.norm(K - K1, 'fro')

print(norm_P_diff, norm_K_diff)


# Monte Carlo simulation with Ns noise sequence samples
Ns = 100
T = 100
x0 = 40 * np.random.randn(n, 1)
x = np.zeros((n, T + 1, Ns))
u = np.zeros((m, T, Ns))
J = np.zeros((T + 1, Ns))
J[0, :] = (x0.T @ x0).flatten()
w = np.sqrt(1) * np.random.randn(n, T, Ns)

for i in range(Ns):
    x[:, 0, i] = x0.flatten()
    for j in range(1, T + 1):
        u[:, j - 1, i] = K @ x[:, j - 1, i]
        x[:, j, i] = (A + B @ K) @ x[:, j - 1, i] + w[:, j - 1, i]
        J[j, i] = x[:, j, i].T @ Q @ x[:, j, i] + u[:, j - 1, i].T @ R @ u[:, j - 1, i]

# Plot sample trajectory
plt.figure(figsize=(12, 6))
plt.step(range(T + 1), x[:, :, 0].T, linewidth=2)
plt.xlabel('Time')
plt.ylabel('States')
plt.title('Sample State Trajectory')
plt.grid(True)
plt.show()

plt.figure(figsize=(12, 6))
plt.step(range(T), u[:, :, 0].T, linewidth=2)
plt.xlabel('Time')
plt.ylabel('Inputs')
plt.title('Sample Input Trajectory')
plt.grid(True)
plt.show()
